enhancements: compound-bundle-constraints (#97)
operator-framework/enhancements: enhancements/compound-bundle-constraints.md

---
title: compound-bundle-constraints
authors:
  - "@estroz"
reviewers:
  - "@joelanford"
  - "@kevinrizza"
  - "@dinhxuanvu"
approvers:
  - "@joelanford"
  - "@kevinrizza"
  - "@dinhxuanvu"
creation-date: 2021-10-14
last-updated: 2021-10-14
status: implementable
see-also:
  - "/enhancements/operator-dependency-resolution.md"
replaces:
superseded-by:
---

# compound-bundle-constraints

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

## Open Questions

1. Eventually I think it would be a good idea to create a general `constraints` property
because, as we add more constraint properties to the top-level bundle config, it becomes
difficult to differentiate what is a property (an intrinsic property of a bundle) and
what is a property that acts as a constraint (a property of a bundle _describing other packages_).
Is this actually a good idea, and is this outside the scope of this EP?

>   - From what I have heard, this is outside the scope of this EP. Would like input though.

Reviewer: IMO, there needs to be _some_ mechanism for OLM to distinguish arbitrary properties from constraints, because it is inevitable that we'll get feature requests that dictate that the resolver doesn't simply ignore them. I think this is a fundamental prerequisite for both this EP and for the generic constraint EP in #91.

version 1.0.0 _and_ a bundle that provides the `EtcdBackup` API in group `etcd.database.coreos.com`
and at version `v1beta2`. The resolver will translate those constraints into a boolean expression
that is evaluated over all installed bundles's properties in the cluster; if both `bar` and `etcd`
bundles with properties satisfying group and/or version constraints are installed, then
`foo.v1.0.0` installation can proceed.

However, what if:
- `foo.v1.0.0` can use multiple versions of the same API, but only one of that group-kind?
- `foo.v1.0.0` can use multiple versions of the same API, but only if another package version is present?
- `foo.v1.0.0` cannot use a particular API version, but can use all others?

These constraints are logical disjunctions and negations in character; the existing
top-level constraints are together a logical conjunction, although not explicit.
Currently the spec does not support specifying such "compound" dependency constraints.

### Goals

- Define logical conjunction, disjunction, and negation properties as dependency constraints, termed "compound constraints".

> - Define nesting of compound constraints constraints.

Reviewer 2: I'm assuming this does not cover 'soft dependencies' where an operator might be installed in ways that would dynamically alter its dependencies?  Is that out of scope for this?

Reviewer 3: I think there might be a typo in this line as well "constraints constraints"

should emit errors so bundles are not installed incorrectly, ex. missing an API.

### Implementation Details/Notes/Constraints

**Note**: see ["Implementation History"](#implementation-history) for proof-of-concept info.

I propose the new properties `olm.all.required`, `olm.any.required`, and `olm.none.required`
to specify conjunction, disjunction, and negation compound constraints, respectively.
Their value type is a list of `olm.package.required`, `olm.gvk.required`, or any
compound constraint listed above.

#### Conjunction and disjunction

These compound constraint types are evaluated following their logical definitions.

This is an example of a conjunctive constraint of two packages and one GVK,
i.e. they must all be satisfied by installed bundles:

```yaml
schema: olm.bundle
name: baz.v1.0.0
properties:
- type: olm.all.required
  value:
  - type: olm.package.required
    value:
      packageName: baz
      versionRange: '>=1.0.0'
  - type: olm.package.required
    value:
      packageName: buf
      versionRange: '>=1.0.0'
  - type: olm.gvk.required
    value:
      group: bazs.example.com
      version: v1

>       kind: Baz

Reviewer: Should `value` be a struct with a string `description` and a slice of sub-`constraints`. I seem to recall some prototype code looking like that.

Author: A description would be useful! I am in favor of changing `value` like that.

        version: v1
        kind: Foo
  - type: olm.all.required
    value:
    - type: olm.package.required
      value:
        packageName: foo
        versionRange: '<1.0.0'
    - type: olm.gvk.required
      value:
        group: foos.example.com
        version: v1beta1
        kind: Foo
```

#### Changes to OLM's resolver

The [resolver package][olm-resolver-primitives] supplies the primitives "and" and "or" already,
and uses them to combine simple GVK and package constraints into an expression.
Compound constraints would simply extend the usage of these primitives in the nested
case, and add the "not" primitive.

### Risks and Mitigations

- With the possibility for nesting comes the increased possibility that two or more nestings
will logically conflict and result in an unsatisfiable expression.
  - There is no quick way to determine whether an expression is unsatisfiable
  other than by using a SAT solver, which the resolver uses internally.
  Compound constraints must be vetted by hand.
- Compound constraint nesting exposes an exponentiation attack vector, where a malicious
bundle author adds deep nesting to a bundle resulting in long resolver pauses.

>   - Tools that work with bundle constraints should limit the depth of nesting.

Author: The POC has a depth limit of 10, which I think is reasonable given the context. I'll state this as a limitation, and mention that this can be increased in the future if deemed necessary

  I recommend the depth limit be 10 levels.
  - Typically bundles are vetted by catalog maintainers, so deep nesting can be caught
  by automation prior to inclusion in a catalog.

## Design Details

### Test Plan

- operator-registry unit and e2e tests.
- operator-lifecycle-manager unit and e2e tests.
  - Both API and resolver unit tests.

### Graduation Criteria

> The alpha feature gate will be removed and this feature will be on permanently by OLM v1.0.0.

Author: This decision was arbitrary. Perhaps v0.(y+2) from release is better, since we want feedback on how well this feature tackles the outlined use cases vs. improvements to the spec that result in breaking changes.

#### Examples

See ["Implementation Details/Notes/Constraints"](#implementation-details-notes-constraints).

##### Dev Preview -> Tech Preview

- Ability to utilize the enhancement end to end
- End user documentation, relative API stability
- Sufficient test coverage
- Gather feedback from users rather than just developers

##### Tech Preview -> GA

- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default

**For non-optional features moving to GA, the graduation criteria must include
end to end tests.**

##### Removing a deprecated feature

- Announce deprecation and support policy of the existing feature
- Deprecate the feature

### Upgrade / Downgrade Strategy

Older versions of OLM will ignore unknown bundle properties, so compound constraints
will be ignored. Starting with the first release containing this feature, the same
holds for all properties not listed as feature gated properties; if the feature gate
is not turned on for some property, OLM will emit an error. Therefore the same bundle
will "work" between OLM versions, even though older versions will install the bundle

> without error by design.

Reviewer: Out-of-scope for this particular EP, but somewhat foundational. If we know we'll eventually want to support `optional` and/or `recommended` constraints, then it seems like it may be better to make a separate top-level constraint field for this enum value. Something like:
```yaml
type: olm.all
necessity: required|recommended|optional
value: { ... }
```

> without error by design.

Author: Moved to version skew. I agree, distinguishing between constraint necessity would be valuable. Worthy of an operator-registry GH issue?

should emit errors so bundles are not installed incorrectly, ex. missing an API.

### Implementation Details/Notes/Constraints

**Note**: see ["Implementation History"](#implementation-history) for proof-of-concept info.

I propose the new properties `olm.all.required`, `olm.any.required`, and `olm.none.required`
to specify conjunction, disjunction, and negation compound constraints, respectively.
Their value type is a struct containing a `description`, and list of `olm.package.required`,
`olm.gvk.required`, or any compound constraint listed above under a `constraints` key.

The implementation of these constraints will primarily live in the OLM codebase,
since they are properties that the resolver understands. Some parsing details
will be implemented in operator-registry for registry components that need to know
about dependencies.

#### Conjunction and disjunction

These compound constraint types are evaluated following their logical definitions.

This is an example of a conjunctive constraint of two packages and one GVK,
i.e. they must all be satisfied by installed bundles:

```yaml
schema: olm.bundle
name: baz.v1.0.0
properties:
- type: olm.all.required
  value:

>     description: All are required for Baz because...

Author: I'm of a similar opinion. For now I'm not going to modify the metaschema, and leave description in value, then update the spec in a follow-up.

        version: v1
        kind: Baz
```

This is an example of a disjunctive constraint of three versions of the same GVK,
i.e. at least one must be satisfied by installed bundles:

```yaml
schema: olm.bundle
name: baz.v1.0.0
properties:
- type: olm.any.required
  value:
    description: Any are required for Baz because...
    constraints:
    - type: olm.gvk.required
      value:
        group: foos.example.com
        version: v1beta1
        kind: Foo
    - type: olm.gvk.required
      value:
        group: foos.example.com
        version: v1beta2
        kind: Foo
    - type: olm.gvk.required
      value:
        group: foos.example.com
        version: v1
        kind: Foo
```

#### Negation

Negation is worth further explanation, since at first glance its semantics
are unclear in this context. The negation is really instructing the resolver
to remove any possible solution that includes a particular GVK, package
at a version, or satisfies some child compound constraint from the result set.

This is an example of a negation constraint of one version of a GVK,
i.e. this GVK cannot satisfy the constraint:

```yaml
schema: olm.bundle
name: baz.v1.0.0
properties:

> - type: olm.none.required

Reviewer 4: this is interesting, but I think it more matches "conflicts" constraints in the resolver than it does negation.

negation would let you say things like `A && B || A && !C` etc

Author: I am unclear on what you mean. Negation here intentionally means "use any GVK/package that satisfies other constraints except these GVKs/packages". Are you saying this is not useful, or just that the implementation of negation in the resolver is "conflicts"?

Upgrading OLM will result in the same resolution results for bundles with top-level
GVK and package constraints.

Downgrading OLM on a cluster with catalogs containing compound constraints will
not cause bundle installation errors, but the resolver will ignore compound
constraint properties. See [version skew strategy](#version-skew-strategy) for details.

### Version Skew Strategy

Older versions of OLM will ignore unknown bundle properties, so compound constraints
will be ignored. Starting with the first release containing this feature, the same
holds for all properties not listed as feature gated properties; if the feature gate
is not turned on for some property, OLM will emit an error. Therefore the same bundle
will "work" between OLM versions, even though older versions will install the bundle
without error by design.

## Implementation History

Proof of concept:
- operator-registry: https://github.com/operator-framework/operator-registry/pull/814
- operator-lifecycle-manager: https://github.com/operator-framework/operator-lifecycle-manager/pull/2418

## Drawbacks

- Runtime penalty for complex boolean expressions in the resolver.

## Alternatives

> None(?)

Reviewer 4: I would be really interested to see alternatives - This proposal does a good job of describing "encoding compound constraints as a constraint" but it doesn't look at any options outside of that.

I just looked at #91 so CEL is on my mind, but if you had a way to reference other constraints you could do things like:

```yaml
constraintA: <constraint>
constraintB: <constraint>
compoundConstraint:
  type: "cel"
  value: "constraintA && constraintB"
```

Author: Ah right this should be an alternative, and it is in-flight already. Thanks for calling this out.

Reviewer: I'm not sure I'm following the suggestion. @ecordell Are you suggesting that constraintA and constraintB are variable delcarations of CEL expressions, which can then be referenced in a compound constraint?

Or are constraintA and constraintB arbitrary non-cel constraints that can then be somehow referenced inside of a cel expression?

Author: I would imagine it's the latter, which I think is not great since you'd be creating ambiguous constraints (i.e. in what order do these get evaluated). Regardless compound constraints could be expressed via an arbitrary constraint expression language, depending on its implementation and what data gets passed to it. I would imagine something more like
```yaml
compoundConstraint:
  type: cel
  value: 'gvk.required("v1", "example.com", "Foo") && package.required("bar", ">=1.0.0")'
```

can be satisfy the constraint.

```yaml
schema: olm.bundle
name: baz.v1.0.0
properties:
- type: olm.any.required
  value:
    description: Required for Baz because...
    constraints:
    - type: olm.all.required
      value:
        description: All are required for Baz because...
        constraints:
        - type: olm.package.required
          value:
            packageName: foo
            versionRange: '>=1.0.0'
        - type: olm.gvk.required
          value:
            group: foos.example.com
            version: v1
            kind: Foo
    - type: olm.all.required
      value:
        description: All are required for Baz because...
        constraints:
        - type: olm.package.required
          value:
            packageName: foo
            versionRange: '<1.0.0'
        - type: olm.gvk.required
          value:
            group: foos.example.com
            version: v1beta1
            kind: Foo
```

The maximum depth of nesting is 10 levels.
This limit can be increased at a later date if necessary.

#### Changes to OLM's resolver

The [resolver package][olm-resolver-primitives] supplies the primitives "and" and "or" already,

> and uses them to combine simple GVK and package constraints into an expression.

Reviewer 3: ```suggestion
and uses them to combine simple GVK and package constraints into a predicate expression.
```

Compound constraints would simply extend the usage of these primitives in the nested
case, and add the "not" primitive.

### Risks and Mitigations

- With the possibility for nesting comes the increased possibility that two or more nestings
will logically conflict and result in an unsatisfiable expression.
  - There is no quick way to determine whether an expression is unsatisfiable
  other than by using a SAT solver, which the resolver uses internally.
  Compound constraints must be vetted by hand.
- Compound constraint nesting exposes an exponentiation attack vector, where a malicious
bundle author adds deep nesting to a bundle resulting in long resolver pauses.

>   - Tools that work with bundle constraints _must_ limit the depth of nesting of 10 levels.

Author: It will be applied, see [the POC](https://github.com/operator-framework/operator-lifecycle-manager/pull/2418/files#diff-28e6ffc774e00e9ab4df1edd63a71768056c1af7905f49da777c185bbf0421c5R849).

I propose the new [`olm.constraint` property][arb-constraints] fields `all`, `any`, and `none`
to specify conjunction, disjunction, and negation compound constraints, respectively.
Their value type is a struct containing a list of any constraint such as `olm.package.required`,
`olm.gvk.required`, or any compound constraint listed above under a `constraints` key.

The implementation of these constraints will primarily live in the OLM codebase,
since they are properties that the resolver understands. Some parsing details
will be implemented in operator-registry for registry components that need to know
about dependencies.

#### Conjunction and disjunction

These compound constraint types are evaluated following their logical definitions.

This is an example of a conjunctive constraint of two packages and one GVK,
i.e. they must all be satisfied by installed bundles:

```yaml
schema: olm.bundle
name: baz.v1.0.0
properties:
- type: olm.constraint
  value:
    message: All are required for Baz because...
    all:
      constraints:
      - type: olm.package.required
        value:
          packageName: baz
          versionRange: '>=1.0.0'

>       - type: olm.package.required
>         value:
>           packageName: buf
>           versionRange: '>=1.0.0'
>       - type: olm.gvk.required
>         value:
>           group: bazs.example.com
>           version: v1
>           kind: Baz

Author: While we're here, I may as well apply `olm.constraint` to GVK/package. Any objections?
```suggestion
      - type: olm.constraint
        value:
          package:
            packageName: buf
            versionRange: '>=1.0.0'
      - type: olm.constraint
        value:
          gvk:
            group: bazs.example.com
            version: v1
            kind: Baz
```

#### Story 5

As a cluster administrator, I should be able to opt-in to resolving compound
constraints in OLM while the feature is alpha. When opted-out, the resolver
should emit errors so bundles are not installed incorrectly, ex. missing an API.

### Implementation Details/Notes/Constraints

**Note**: see ["Implementation History"](#implementation-history) for proof-of-concept info.

I propose the new [`olm.constraint` property][arb-constraints] fields `all`, `any`, and `none`
to specify conjunction, disjunction, and negation compound constraints, respectively.
Their value type is a struct containing a list of constraints such as `olm.package.required`,
`olm.gvk.required`, or any compound constraint listed above under a `constraints` key.

I also propose that `olm.gvk.required` and `olm.package.required` are redefined as
`gvk` and `package` fields to align with other properties under `olm.constraint`.
The only other difference between existing and new GVK/package constraint schemas
is the new schemas cannot be specified as top-level, non-compound fields; they must
always be housed under one of `all`, `any`, or `none`. Given this difference,

> OLM will only interpret one set of schemas or the other as inputs to the resolver:

Reviewer: Shouldn't it also be possible to have something like this:

```yaml
type: olm.constraint
value:
  gvk:
    group: example.com
    version: v1
    kind: MyApp
```

Basically you can only use `gvk` and `package` within an `olm.constraint` value. But I don't think it should be required that these two types are nested within a compound constraint.

Author: My thought process was that top-level constraints are always a conjunction, so why not make it explicit as `all`. I am ok relaxing this given most users will probably not need compound constraints.

1. If at least one `olm.constraint` is present, interpret only `olm.constraint` properties

> and ignore all other constraint properties.

Reviewer: So anyone wanting use to a `cel` expression from #91 would need to also translate their `olm.gvk.required` and `olm.package.required` properties to be `olm.constraint` properties with `gvk` and `package` sub-fields?

I'm trying to understand the motivation for new versions of OLM ignoring the existing root level `olm.package.required` and `olm.gvk.required`.

---
title: compound-bundle-constraints
authors:
  - "@estroz"
reviewers:
  - "@joelanford"
  - "@kevinrizza"
  - "@dinhxuanvu"
approvers:
  - "@joelanford"
  - "@kevinrizza"
  - "@dinhxuanvu"
creation-date: 2021-10-14
last-updated: 2021-11-18
status: implementable
see-also:
  - "/enhancements/operator-dependency-resolution.md"
replaces:
superseded-by:
---

# compound-bundle-constraints

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

## Open Questions

1. Eventually I think it would be a good idea to create a general `constraints` property
because, as we add more constraint properties to the top-level bundle config, it becomes
difficult to differentiate what is a property (an intrinsic property of a bundle) and
what is a property that acts as a constraint (a property of a bundle _describing other packages_).
Is this actually a good idea, and is this outside the scope of this EP?

>   - This has been written into the `olm.constraints` spec in [#91][arb-constraints] and used here.

Author: will do.

These constraints are either `olm.gvk.required`, `olm.package.required`,
or an [`olm.constraint` clause][arb-constraints]. After the resolver completes evaluation,
it either returns a non-satisfiability error or the "best" choice of constraint bundle to install.
Any properties not recognized by the resolver are ignored, whether constraints or otherwise.

## Motivation

The existing constraints `olm.package.required` and `olm.gvk.required` are the only constraints
required for most use cases. They are specified as follows:

```yaml
---
schema: olm.bundle
name: foo.v1.0.0
package: foo
properties:
- type: olm.package.required
  value:
    name: bar
    versionRange: '>=1.0.0'
- type: olm.gvk.required
  value:
    group: etcd.database.coreos.com
    kind: EtcdBackup
    version: v1beta2
```

The above bundle `foo.v1.0.0` requires that both a bundle in package `bar` that is of at least
version 1.0.0 _and_ a bundle that provides the `EtcdBackup` API in group `etcd.database.coreos.com`
and at version `v1beta2`. The resolver will translate those constraints into a boolean expression

> that is evaluated over all installed bundles's properties in the cluster; if both `bar` and `etcd`

Reviewer 3: ```suggestion
that is evaluated over all installed bundles' properties in the cluster; if both `bar` and `etcd`
```

Currently the spec does not support specifying such "compound" dependency constraints.

### Goals

- Define logical conjunction, disjunction, and negation properties as dependency constraints, termed "compound constraints".
- Define nesting of compound constraints.
- Describe how compound constraints are evaluated.
- Describe what needs to change in OLM's resolver to support compound constraints.
- Plan for implementation under an alpha feature gate, and a timeline for GA.

### Non-Goals

- Change the existing constraint definitions.
- Break bundle spec or OLM backwards-compatibility in any way.

## Proposal

### User Stories

#### Story 1

As an operator author, I should be able to specify "any of these APIs"
or "none of these APIs" dependencies as bundle properties, and "all of these APIs" explicitly.

#### Story 2

As an operator author, I should be able to nest any compound constraint(s)
within a parent compound constraint.

#### Story 3

As an operator author, I should be able to build operator bundles that "work"
with both older and newer versions of OLM. I understand that "work" might mean
different things depending on OLM version; obviously older OLM versions cannot

> compound constraints using any method other than conjunction.

Author: It is, will remove.

        package:
          name: bar
          versionRange: '>=1.0.0'
      - message: GVK Buf/v1 is needed for...
        gvk:
          group: bufs.example.com
          version: v1
          kind: Buf
```

This is an example of a disjunctive constraint of three versions of the same GVK,
i.e. at least one must be satisfied by installed bundles:

```yaml
schema: olm.bundle
name: baz.v1.0.0
properties:
- type: olm.constraint
  value:
    message: Any are required for Baz because...
    any:
      constraints:
      - gvk:
          group: foos.example.com
          version: v1beta1
          kind: Foo
      - gvk:
          group: foos.example.com
          version: v1beta2
          kind: Foo
      - gvk:
          group: foos.example.com
          version: v1
          kind: Foo
```

#### Negation

Negation is worth further explanation, since at first glance its semantics
are unclear in this context. The negation is really instructing the resolver
to remove any possible solution that includes a particular GVK, package
at a version, or satisfies some child compound constraint from the result set.

This is an example of a negation constraint of one version of a GVK,

> i.e. this GVK cannot satisfy the constraint:

Author: Yes.

- type: olm.constraint
  value:
    message: Cannot be required for Baz because...
    none:
      constraints:
      - gvk:
          group: foos.example.com
          version: v1alpha1
          kind: Foo
```

#### Nested compound constraints

A nested compound constraint, one that contains at least one child compound constraint
along with zero or more simple constraints, is evaluated from the bottom up following
the procedures described for each above.

This is an example of a disjunction of conjunctions, where one, the other, or both
can be satisfy the constraint.

```yaml
schema: olm.bundle
name: baz.v1.0.0
properties:
- type: olm.constraint
  value:
    message: Required for Baz because...
    any:
      constraints:
      - all:
          constraints:
          - package:
              name: foo
              versionRange: '>=1.0.0'
          - gvk:
              group: foos.example.com
              version: v1
              kind: Foo
      - all:
          constraints:
          - package:
              name: foo
              versionRange: '<1.0.0'
          - gvk:
              group: foos.example.com
              version: v1beta1
              kind: Foo
```

> The maximum raw size of an `olm.constraint` is 64KB to limit resource exhaustion attacks.

Reviewer 3: Nice!
